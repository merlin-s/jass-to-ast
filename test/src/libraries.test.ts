import { parseSnapshot } from "../util";

it(
	"comments",
	parseSnapshot(`
        library A // foo
            // bar
        endlibrary // baz
    `),
);

describe("requires", () => {
	it(
		"none",
		parseSnapshot(`
            library A
            endlibrary
        `),
	);

	it(
		"one",
		parseSnapshot(`
            library B requires A
            endlibrary
        `),
	);

	it(
		"two",
		parseSnapshot(`
            library C requires A, B
            endlibrary
        `),
	);

	it(
		"optionals",
		parseSnapshot(`
            library E requires A, optional B, optional C, D
            endlibrary
        `),
	);
});

describe("globals", () => {
	it(
		"simple",
		parseSnapshot(`
            library A

                globals
                    gamecache cache
                endglobals

            endlibrary
        `),
	);

	it(
		"scoped",
		parseSnapshot(`
            library A
                globals
                    integer globalInt
                    private integer privInt
                    public integer pubInt
                    private constant integer privConstInt
                endglobals
            endlibrary
        `),
	);
});

describe("functions", () => {
	it(
		"general",
		parseSnapshot(`
            library B
                function Bfun takes nothing returns integer
                    local integer i = 0
                    set i = i + 1
                    return i
                endfunction
            endlibrary
        `),
	);

	it(
		"scoped",
		parseSnapshot(`
            library A
                function GlobalFunc takes nothing returns integer
                endfunction

                private function PrivFunc takes nothing returns integer
                endfunction

                public function PubFunc takes nothing returns integer
                endfunction
            endlibrary
        `),
	);
});

it(
	"initializer",
	parseSnapshot(`
        library A initializer Init
            function Init takes nothing returns integer
            endfunction
        endlibrary
    `),
);

it(
	"scopes",
	parseSnapshot(`
        scope A
            private function Init takes nothing returns integer
            endfunction
        endscope
    `),
);

it(
	"modules",
	parseSnapshot(`
    module A
        private module B
        endmodule
    endmodule
`),
);

it(
	"keywords",
	parseSnapshot(`
        module Foo
            private keyword B // test
        endmodule
    `),
);
