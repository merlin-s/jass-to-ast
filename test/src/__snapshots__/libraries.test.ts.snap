// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`comments 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        Comment {
          "value": " bar",
        },
      ],
      "comment": Comment {
        "value": " foo",
      },
      "endComment": Comment {
        "value": " baz",
      },
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;

exports[`functions general 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        JASSFunction {
          "access": "global",
          "name": Name {
            "value": "Bfun",
          },
          "returns": "integer",
          "statements": Statements {
            "body": Array [
              Variable {
                "access": "local",
                "name": Name {
                  "value": "i",
                },
                "type": "integer",
                "value": 0,
              },
              JASSSet {
                "name": Name {
                  "value": "i",
                },
                "value": BinaryOp {
                  "left": Name {
                    "value": "i",
                  },
                  "operator": "+",
                  "right": 1,
                },
              },
              Return {
                "data": Array [
                  Name {
                    "value": "i",
                  },
                ],
              },
            ],
          },
        },
      ],
      "name": Name {
        "value": "B",
      },
    },
  ],
}
`;

exports[`functions scoped 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        JASSFunction {
          "access": "global",
          "name": Name {
            "value": "GlobalFunc",
          },
          "returns": "integer",
        },
        EmptyLine {},
        JASSFunction {
          "access": "private",
          "name": Name {
            "value": "PrivFunc",
          },
          "returns": "integer",
        },
        EmptyLine {},
        JASSFunction {
          "access": "public",
          "name": Name {
            "value": "PubFunc",
          },
          "returns": "integer",
        },
      ],
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;

exports[`globals scoped 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        Globals {
          "globals": Statements {
            "body": Array [
              Variable {
                "access": "global",
                "name": Name {
                  "value": "globalInt",
                },
                "type": "integer",
              },
              Variable {
                "access": "private",
                "name": Name {
                  "value": "privInt",
                },
                "type": "integer",
              },
              Variable {
                "access": "public",
                "name": Name {
                  "value": "pubInt",
                },
                "type": "integer",
              },
              Variable {
                "access": "private",
                "constant": true,
                "name": Name {
                  "value": "privConstInt",
                },
                "type": "integer",
              },
            ],
          },
        },
      ],
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;

exports[`globals simple 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        EmptyLine {},
        Globals {
          "globals": Statements {
            "body": Array [
              Variable {
                "access": "global",
                "name": Name {
                  "value": "cache",
                },
                "type": Name {
                  "value": "gamecache",
                },
              },
            ],
          },
        },
        EmptyLine {},
      ],
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;

exports[`initializer 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        JASSFunction {
          "access": "global",
          "name": Name {
            "value": "Init",
          },
          "returns": "integer",
        },
      ],
      "initializer": Name {
        "value": "Init",
      },
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;

exports[`requires none 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [],
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;

exports[`requires one 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [],
      "name": Name {
        "value": "B",
      },
      "requires": Array [
        Requirement {
          "name": Name {
            "value": "A",
          },
          "optional": false,
        },
      ],
    },
  ],
}
`;

exports[`requires optionals 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [],
      "name": Name {
        "value": "E",
      },
      "requires": Array [
        Requirement {
          "name": Name {
            "value": "A",
          },
          "optional": false,
        },
        Requirement {
          "name": Name {
            "value": "B",
          },
          "optional": true,
        },
        Requirement {
          "name": Name {
            "value": "C",
          },
          "optional": true,
        },
        Requirement {
          "name": Name {
            "value": "D",
          },
          "optional": false,
        },
      ],
    },
  ],
}
`;

exports[`requires two 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [],
      "name": Name {
        "value": "C",
      },
      "requires": Array [
        Requirement {
          "name": Name {
            "value": "A",
          },
          "optional": false,
        },
        Requirement {
          "name": Name {
            "value": "B",
          },
          "optional": false,
        },
      ],
    },
  ],
}
`;

exports[`scopes 1`] = `
Program {
  "body": Array [
    Library {
      "blocks": Array [
        JASSFunction {
          "access": "private",
          "name": Name {
            "value": "Init",
          },
          "returns": "integer",
        },
      ],
      "name": Name {
        "value": "A",
      },
    },
  ],
}
`;
